### ISSUE TITLE
OpenGL Overlay causes massive "Late Start"s

### ISSUE THREAD

**Issue by huheyomservice** [ValveSoftware/openvr#1598]
i have written a small plugin for a game that draws an overlay in its own thread.
i created a window and draw with opengl inside it, copy it to a texture and then send it to the overlay with setoverlaytexture();

it does work how it should, but as soon as the overlay is visible, i am getting huge amount of late starts.
debugging the code shows barely any time lost during one render loop.
i have also added a 11ms (later 60ms) Sleep, which had no (positive) effect on the endresult.

my assumption was/is that if i dont send a new texture to the overlay, than it will just keep using the old on.
or do i have to keep sending new ones before the next frame ?

aka: do i need to reduce/remove the sleep altogether in order to get a smooth experience/ is the sleep causing those problems.

currently at work, so i am just asking prematurely to have something to try when i get home.

PS: the original renderthread from the game is not being touched and running seperately, so should (in theory) have no issues running.

PSS: within my own application there is barely anything happening, just a simple rectangle being drawn but even that is causing issues.

**Comment by huheyomservice** [comment]
i have done a bit more research on this.

this only seems to happen if i am injecting the dll once the game is already in a playing state aka stuff is being submitted to the compositor.

if i inject right at the start in the main menu, than my overlay seems to render fine without any latestarts but the issue then is, that non of the 3d world is being rendered. or the 3d world doesnt seem to update anymore, like its still updating the frames but the transform/pose of my hmd doesnt seem to be applied, it always looks in one direction, if i look 90 to the side than i am looking into blackness.

very weird

**Comment by huheyomservice** [comment]
any idea if having some textures as d3dx11 and some textures as opengl can cause an issue?

the game seems to set the overlay texture for its own overlay as a d3d11 texture/resource, while my overlay uses a opengl texture (they are 2 seperate overlays)

additionally i have to add that the game uses a really old openvr_api version
if i recall correctly its version 0.9.12 or something like that

**Comment by Rectus** [comment]
Not fully sure, but it seems SteamVR is tracking application state per process, and will set the application type from the fist call to VR_Init(). There doesn't seem to be any documentation on if overlays are supported from scene applications, but I'm guessing not. You would probably need to start the overlay as a separate process if that is the case.

**Comment by huheyomservice** [comment]
> process

hey thanks for your reply.
the program itself has both a scene and 2 overlays by default (its shipped that way) and it works fine

but when creating the third overlay (mine) there seems to be alot of trouble.
i have also hooked the function that calls all the relevant overlay functions from the game, so the timing is also perfect.

additional info: i am not calling vr_init() in my dll, since vr is already inited and things like vr::VROverlay() are all returning positive.

**Comment by Rectus** [comment]
In that case it sounds like the OpenGL textures could be the cause. There is no explicit way of telling it what render API your are using, so it might be autodetecting it based on the textures sent to it, and constantly reinitializing the internal DirectX and OpenGL interfaces.

**Comment by masterlooser** [comment]
> In that case it sounds like the OpenGL textures could be the cause. There is no explicit way of telling it what render API your are using, so it might be autodetecting it based on the textures sent to it, and constantly reinitializing the internal DirectX and OpenGL interfaces.

the weird thing is that all overlays are rendering fine. its either that the 3D world breaks or it gets really choppy FPS wise.

ill try to hook the d3d11 device etc from the game and see if i can make a d3d11 texture.

if that fails too, than ill just hook literally everything in the game and call all the functions the game is calling to create an overlay 😂 
if that fails too, i either give up for figure out how the game renders on its own overlay/texture and use "engine"-drawing functions

**Comment by huheyomservice** [comment]
so as i said, as "masterlooser", i have hooked d3d11 of the game and draw my stuff to a d3d11 texture.
and i call setoverlaytexture inside the hooked present call, which works flawless, i can now see all overlays and the 3D world, without any latestarts.

additionally i grabbed the games vroverlay function, that way i can call the virtuals without the need for the openvr.h/openvr_api.lib, which is confinient since my dll injects before the game has loaded the openvr_api.dll, also the game can run without vr, which wouldnt allow me to inject the DLL

now its compatible with both the vr and non vr version.

thanks everyone who tried helping.

TLDR: dont mix opengl and d3d11 i guess xD
